x/evidence/internal/keeper/infraction.go

```diff
@@ -0,0 +1,104 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/cosmos-sdk/x/evidence/internal/types"
+)
+
+// HandleDoubleSign implements an equivocation evidence handler. Assuming the
+// evidence is valid, the validator committing the misbehavior will be slashed,
+// jailed and tombstoned. Once tombstoned, the validator will not be able to
+// recover. Note, the evidence contains the block time and height at the time of
+// the equivocation.
+//
+// The evidence is considered invalid if:
+// - the evidence is too old
+// - the validator is unbonded or does not exist
+// - the signing info does not exist (will panic)
+// - is already tombstoned
+//
+// TODO: Some of the invalid constraints listed above may need to be reconsidered
+// in the case of a lunatic attack.
+func (k Keeper) HandleDoubleSign(ctx sdk.Context, evidence types.Equivocation) {
+	logger := k.Logger(ctx)
+	consAddr := evidence.GetConsensusAddress()
+	infractionHeight := evidence.GetHeight()
+
+	// calculate the age of the evidence
+	blockTime := ctx.BlockHeader().Time
+	age := blockTime.Sub(evidence.GetTime())
+
+	if _, err := k.slashingKeeper.GetPubkey(ctx, consAddr.Bytes()); err != nil {
+		// Ignore evidence that cannot be handled.
+		//
+		// NOTE: We used to panic with:
+		// `panic(fmt.Sprintf("Validator consensus-address %v not found", consAddr))`,
+		// but this couples the expectations of the app to both Tendermint and
+		// the simulator.  Both are expected to provide the full range of
+		// allowable but none of the disallowed evidence types.  Instead of
+		// getting this coordination right, it is easier to relax the
+		// constraints and ignore evidence that cannot be handled.
+		return
+	}
+
+	// reject evidence if the double-sign is too old
+	if age > k.MaxEvidenceAge(ctx) {
```

Comment:
q: shouldn't every evidence type define its own evidence age instead?

Reply:
I don't think so because all evidence must be within the unbonding period. Perhaps certain evidence can have even smaller valid periods (e.g. 1/2 unbonding period), but I don't see any immediate use-cases or evidence types that need this.